localstack/services/s3/storage.py

```diff
@@ -0,0 +1,282 @@
+import abc
+from collections import defaultdict
+from shutil import rmtree
+from tempfile import SpooledTemporaryFile, mkdtemp
+from threading import RLock
+from typing import Iterator, TypedDict
+
+from readerwriterlock import rwlock
+
+from localstack.aws.api.s3 import (
+    BucketName,
+    MultipartUploadId,
+    ObjectKey,
+    ObjectVersionId,
+    PartNumber,
+)
+from localstack.services.s3.constants import S3_CHUNK_SIZE
+
+# max file size for S3 objects kept in memory (500 KB by default)
+# TODO: make it configurable
+S3_MAX_FILE_SIZE_BYTES = 512 * 1024
+
+
+class BaseStorageBackend(abc.ABC):
+    """
+    Base class abstraction for S3 Storage Backend. This allows decoupling between S3 metadata and S3 file objects
+    """
+
+    def create_bucket_directory(self, bucket_name: BucketName):
+        raise NotImplementedError
+
+    def delete_bucket_directory(self, bucket_name: BucketName):
+        raise NotImplementedError
+
+    def create_upload_directory(self, bucket_name: BucketName, upload_id: MultipartUploadId):
+        raise NotImplementedError
+
+    def delete_upload_directory(self, bucket_name: BucketName, upload_id: MultipartUploadId):
+        raise NotImplementedError
```

Comment:
Following up on the StorageBackend abstraction discussion: I'm also wondering whether the BaseStorageBackend shouldn't have fewer but more high-level methods. Like the creation of directories and such, shouldn't that be hidden from the S3Provider?

Reply:
I thought about it too, that it should be hidden, that was how I first wrote it. But then, if it's hidden from the provider, when do we create a directory? The first time we create a file object in the "bucket"? Then we need to check every time if the directory already exists. Following the "lifecycle" of the provider, which to me is the "orchestrator" of both Metadata (KeyStore, S3Buckets and such) and the StorageBackend. The issue as you pointed out is the synchronisation between both. We could need some kind of lock to be sure they're in sync, but already following the approach of always doing the updates in the metadata in the right order, but it need to be conscious. This is, in my opinion, the biggest flaw of the split between metadata/assets.